internal/api/oidc/userinfo.go

```diff
@@ -0,0 +1,276 @@
+			PreferredUsername: user.PreferredLoginName,
+		}
+	}
+	if machine := user.Machine; machine != nil {
+		return oidc.UserInfoProfile{
+			Name:              machine.Name,
+			UpdatedAt:         oidc.FromTime(user.ChangeDate),
+			PreferredUsername: user.PreferredLoginName,
+		}
+	}
+	return oidc.UserInfoProfile{}
+}
+
+func userInfoPhoneToOIDC(user *query.User) oidc.UserInfoPhone {
+	if human := user.Human; human != nil {
+		return oidc.UserInfoPhone{
+			PhoneNumber:         string(human.Phone),
+			PhoneNumberVerified: human.IsPhoneVerified,
+		}
+	}
+	return oidc.UserInfoPhone{}
+}
+
+func setUserInfoMetadata(metadata []query.UserMetadata, out *oidc.UserInfo) {
+	if len(metadata) == 0 {
+		return
+	}
+	mdmap := make(map[string]string, len(metadata))
+	for _, md := range metadata {
+		mdmap[md.Key] = base64.RawURLEncoding.EncodeToString(md.Value)
+	}
+	out.AppendClaims(ClaimUserMetaData, mdmap)
+}
+
+func setUserInfoOrgClaims(user *query.OIDCUserInfo, out *oidc.UserInfo) {
+	if org := user.Org; org != nil {
+		out.AppendClaims(ClaimResourceOwner+"id", org.ID)
+		out.AppendClaims(ClaimResourceOwner+"name", org.Name)
+		out.AppendClaims(ClaimResourceOwner+"primary_domain", org.PrimaryDomain)
+	}
+}
+
+func setUserInfoRoleClaims(userInfo *oidc.UserInfo, roles *projectsRoles) {
+	if roles != nil && len(roles.projects) > 0 {
+		if roles, ok := roles.projects[roles.requestProjectID]; ok {
+			userInfo.AppendClaims(ClaimProjectRoles, roles)
+		}
+		for projectID, roles := range roles.projects {
+			userInfo.AppendClaims(fmt.Sprintf(ClaimProjectRolesFormat, projectID), roles)
+		}
+	}
+}
+
+func (s *Server) userinfoFlows(ctx context.Context, qu *query.OIDCUserInfo, userInfo *oidc.UserInfo) error {
+	queriedActions, err := s.query.GetActiveActionsByFlowAndTriggerType(ctx, domain.FlowTypeCustomiseToken, domain.TriggerTypePreUserinfoCreation, qu.User.ResourceOwner, false)
+	if err != nil {
+		return err
+	}
+
+	ctxFields := actions.SetContextFields(
+		actions.SetFields("v1",
+			actions.SetFields("claims", userinfoClaims(userInfo)),
+			actions.SetFields("getUser", func(c *actions.FieldConfig) interface{} {
+				return func(call goja.FunctionCall) goja.Value {
+					return object.UserFromQuery(c, qu.User)
+				}
+			}),
+			actions.SetFields("user",
+				actions.SetFields("getMetadata", func(c *actions.FieldConfig) interface{} {
+					return func(goja.FunctionCall) goja.Value {
+						return object.UserMetadataListFromSlice(c, qu.Metadata)
+					}
+				}),
+				actions.SetFields("grants", func(c *actions.FieldConfig) interface{} {
+					return object.UserGrantsFromSlice(c, qu.UserGrants)
+				}),
+			),
+		),
+	)
+
+	for _, action := range queriedActions {
+		actionCtx, cancel := context.WithTimeout(ctx, action.Timeout())
+		claimLogs := []string{}
+
+		apiFields := actions.WithAPIFields(
+			actions.SetFields("v1",
+				actions.SetFields("userinfo",
+					actions.SetFields("setClaim", func(key string, value interface{}) {
```

Comment:
Isn't userinfo and claims the same here?

Reply:
Yes, I believe you are right. This code is copy-pasted from the origninal implementation, which will be cleaned up once we port unserinfo to the new implementation:

https://github.com/zitadel/zitadel/blob/2e2214339ed3925dbdc52e44eeea631ab3562f32/internal/api/oidc/client.go#L509-L532

However, we do not want to change any output behavior of the introspection endpoint, even if the original output somehow seems wrong. If we remove the `userInfo` or `claims` fields, any action relying on either of them would no longer work. They are also both part of the [documentation](https://zitadel.com/docs/apis/actions/complement-token).